File: CodeRelayAndroid/core-net/src/main/kotlin/relay/net/PromptOptimizer.kt
package relay.net

/**
 * Result of an `optimize_prompt` RPC, ported from `PromptOptimizer.swift`.
 * Unknown statuses from a newer relay collapse into [Failed] so they can never
 * crash an older client.
 */
sealed interface OptimizeOutcome {
    /**
     * The server rewrote the draft in place. [original] is what was on the
     * input line before; null only against a relay that omits it, in which
     * case Undo cannot be offered.
     */
    data class Ok(val original: String?) : OptimizeOutcome
    data object NoDraft : OptimizeOutcome
    data object Passthrough : OptimizeOutcome
    data object Unconfigured : OptimizeOutcome
    /** [message] is the server's text, or [OptimizerStrings.COULD_NOT_REWRITE] when it sent none. */
    data class Failed(val message: String) : OptimizeOutcome
}

/** Result of a `replace_prompt` (Undo) RPC. */
sealed interface ReplaceOutcome {
    data object Ok : ReplaceOutcome
    data class Failed(val message: String) : ReplaceOutcome
}

/**
 * Every user-facing optimizer string, byte-exact with `OptimizerStrings` in
 * `Sources/CodeRelayClient/PromptOptimizer.swift` (spec §7.1, §9, §10). Defined
 * once; call sites must reference these, never retype them.
 */
object OptimizerStrings {
    const val CONFIG_HINT = "Enable on the relay: claude-relay config set promptOptimizerEnabled true"
    const val UPDATE_RELAY_HINT = "Update the relay to use the prompt optimizer"
    const val NO_DRAFT = "Type or dictate a prompt first"
    const val PASSTHROUGH = "Nothing to optimize"
    const val COULD_NOT_REWRITE = "Optimizer could not rewrite this prompt"
    const val OPTIMIZED = "Optimized"
    const val UNDO = "Undo"
    const val WAND_LABEL = "Optimize Prompt"
    const val SHARE_SCREEN_TOGGLE = "Share terminal screen with the optimizer"
    const val SHARE_SCREEN_FOOTER =
        "With this on, the relay also sends the last 40 lines of the terminal screen to the optimizer model. " +
            "The draft and its working directory are always sent."
}

/** Wire-level facts the wand's availability gate is derived from (spec §6, §7.1). */
object PromptOptimizerProtocol {
    /**
     * The first protocol version whose `auth_success` carries `capabilities`.
     * Deliberately a literal and NOT [ProtocolVersions.CURRENT]: the gate must
     * not move when the client's own version does.
     */
    const val MIN_PROTOCOL_VERSION = 2

    /** The capability string the relay advertises when the optimizer is usable. */
    const val CAPABILITY = "prompt_optimizer"
}

File: Sources/CodeRelayClient/PromptOptimizer.swift
import Foundation

/// Client-side view of `optimize_prompt_result` (spec §6). Unknown statuses
/// collapse into `.failed` so a newer server can never crash an older client.
public enum OptimizeOutcome: Equatable, Sendable {
    /// The relay replaced the draft. `original` is what it replaced, for Undo;
    /// the relay may omit it (nil) when it was not tracking the draft.
    case ok(original: String?)
    case noDraft
    case passthrough
    case unconfigured
    case failed(message: String)
}

/// Client-side view of `replace_prompt_result` (spec §6).
public enum ReplaceOutcome: Equatable, Sendable {
    case ok
    case failed(message: String)
}

/// Every user-facing string of the prompt-optimizer feature, byte-exact per
/// spec §7.1 / §9 / §10. Both apps read these; nothing retypes them.
public enum OptimizerStrings {
    public static let configHint = "Enable on the relay: claude-relay config set promptOptimizerEnabled true"
    public static let updateRelayHint = "Update the relay to use the prompt optimizer"
    public static let noDraft = "Type or dictate a prompt first"
    public static let passthrough = "Nothing to optimize"
    public static let couldNotRewrite = "Optimizer could not rewrite this prompt"
    public static let optimized = "Optimized"
    public static let undo = "Undo"
    public static let wandLabel = "Optimize Prompt"
    public static let shareScreenToggle = "Share terminal screen with the optimizer"
    public static let shareScreenFooter = "With this on, the relay also sends the last 40 lines of the terminal screen to the optimizer model. The draft and its working directory are always sent."
}

public extension Notification.Name {
    /// Posted by the hardware-keyboard shortcut (iOS `RelayTerminalView`,
    /// macOS `RecordingShortcutMonitor`); `WandButton` observes it and taps
    /// itself. Replaces the removed `toggleSpeechRecording`.
    static let optimizePromptShortcut = Notification.Name("optimizePromptShortcut")
}
